amgcl/detail/qr.hpp

```diff
@@ -231,15 +263,63 @@ class QR {
             return q[i*row_stride + j*col_stride];
         }
 
+        // Solve over- or underdetermined system Ax = rhs.
+        // Calls compute() on the appropriate version of A (either transposed
+        // or not).
+        void solve(int rows, int cols, value_type *A, value_type *rhs, value_type *x) {
+
+            const int row_stride = (Order == row_major ? nmax : 1);
+            const int col_stride = (Order == row_major ? 1 : m);
+
+            if (rows >= cols) {
+
+                compute(rows, cols, row_stride, col_stride, A);
+
+                for(int i = 0, ii = 0; i < m; ++i, ii += row_stride + col_stride)
+                    apply_reflector(m-i, 1, r+ii, row_stride, math::adjoint(tau[i]), rhs+i, 1, 1);
+
+                std::copy(rhs, rhs+k, x);
+
+                for(int i = n; i --> 0; ) {
+                    value_type rii = r[i*(row_stride+col_stride)];
+                    if (math::is_zero(rii)) continue;
+                    x[i] = math::inverse(rii) * x[i];
+
+                    for(int j = 0, ja = 0; j < i; ++j, ja += row_stride)
+                        x[j] -= r[ja+i*col_stride] * x[i];
+                }
+
+            } else {
+
+                compute(cols, rows, col_stride, row_stride, A);
+
+                for(int i = 0; i < n; ++i) {
+                    value_type rii = math::adjoint(r[i*(row_stride+col_stride)]);
+                    if (math::is_zero(rii)) continue;
+
+                    for(int j = 0, ja = 0; j < i; ++j, ja += row_stride)
```

Comment:
And this loop should go from `i+1` to `n-1`, since we are solving lower triangular system.